workflows/prognostic_c48_run/runtime/monitor.py

```diff
@@ -0,0 +1,107 @@
+import logging
+from typing import (
+    Callable,
+    Iterable,
+    Set,
+)
+import vcm
+from runtime.types import Diagnostics, State
+
+from .names import DELP
+
+logger = logging.getLogger(__name__)
+
+
+class Monitor:
+    """Utility class for monitoring changes to a state dictionary and returning
+    the outputs as tendencies
+    """
+
+    def __init__(self, variables: Iterable[str], state: State, timestep: float):
+        """
+
+        Args:
+            variables: list of variables with names like
+                `tendency_of_{variable}_due_to_{name}`. Used to infer the variables
+                to be monitored.
+            state: The mutable object to monitor
+            timestep: the length of the timestep used to compute the tendency
+        """
+        # need to consume variables into set to use more than once
+        var_set = set(variables)
+        self.tendency_variables = filter_tendency(var_set)
+        self.storage_variables = filter_storage(var_set)
+        self._state = state
+        self.timestep = timestep
+
+    def __call__(
+        self, name: str, func: Callable[[], Diagnostics],
+    ) -> Callable[[], Diagnostics]:
+        """Decorator to add tendency monitoring to an update function
+
+        This will add the following diagnostics:
+        - `tendency_of_{variable}_due_to_{name}`
+        - `storage_of_{variable}_path_due_to_{name}`. A pressure-integrated version
+        of the above
+        - `storage_of_mass_due_to_{name}`, the total mass tendency in Pa/s.
+
+        Args:
+            name: the name to tag the tendency diagnostics with
+            func: a stepping function
```

Comment:
```suggestion
            func: a stepping function which modifies the `state` this object is monitoring
```

Reply:
Given this tight coupling between `func` and `state`, should `state` be passed at call time instead of on init? Otherwise this class is still tightly coupled to the owning scope (you need to track the relation between the Monitor internal `state` instance and the scope it's being used in).

Reply:
possibly, but that would be a backwards incompatible change to `TimeLoop.monitor`.  In practice, `state` is a singleton anyway since it just wraps `fv3gfs.get/set_state`.

Reply:
Should `monitor` be public API? `git grep monitor .` in fv3net root shows it's only used privately.

Reply:
The only "public API" provided by the prog run is `runfv3` IMO.

Reply:
Then it should be fine to make a backwards-incompatible change to an internal function?

Reply:
For sure...if we think that change is helpful and in-scope.